src/main/java/org/osc/controller/nsfc/api/NeutronSfcSdnRedirectionApi.java

```diff
@@ -246,25 +248,84 @@ public InspectionHookElement getInspectionHook(String inspectionHookId) throws E
 
     // SFC methods
     @Override
-    public NetworkElement registerNetworkElement(List<NetworkElement> inspectedPorts) throws Exception {
-        return null;
+    public NetworkElement registerNetworkElement(List<NetworkElement> inspectionPorts) throws Exception {
+       
+        ServiceFunctionChainEntity sfc = new ServiceFunctionChainEntity();
+        //check for null or empty list
+        this.utils.throwExceptionIfNullOrEmptyNetworElementList(inspectionPorts, "PortPairGroup memeber list");
+     
+        this.utils.addPortPairGroupToServiceFunction(inspectionPorts, sfc);
+        return this.txControl.required(() -> {
+            this.em.merge(sfc);
+            return this.utils.sendServiceFunctionChainId(sfc);
+        });
+       
     }
 
     @Override
-    public NetworkElement updateNetworkElement(NetworkElement portGroup, List<NetworkElement> inspectedPorts)
+    public NetworkElement updateNetworkElement(NetworkElement portGroup, List<NetworkElement> inspectionPorts)
             throws Exception {
-        // no-op
-        return null;
+        
+        this.utils.throwExceptionIfNullElementAndParentId(portGroup, "PortPairGroup ServiceFunctionChain Id");
+        this.utils.throwExceptionIfNullOrEmptyNetworElementList(inspectionPorts, "PortPairGroup update memeber list");
+       
+        return this.txControl.required(() -> {
+            ServiceFunctionChainEntity sfc = this.utils.findBySfcId(portGroup.getParentId());
+            
+            if(sfc == null) {
+                throw new IllegalStateException(String.format("Failed to Update serviceFunctionChain : %s ",portGroup.getParentId()
+                         + "\n" + "Reason : Unable to find serviceFunctionChain" , portGroup.getParentId()));
+            }
+            sfc.getPortPairGroups().clear();
+            this.utils.addPortPairGroupToServiceFunction(inspectionPorts, sfc);
+            this.em.merge(sfc);
+            return this.utils.sendServiceFunctionChainId(sfc);
+        });   
     }
 
     @Override
     public void deleteNetworkElement(NetworkElement portGroupId) throws Exception {
-        // no-op
+        this.utils.throwExceptionIfNullElementAndId(portGroupId, "PortPairGroupId");
+        this.utils.throwExceptionIfNullElementAndParentId(portGroupId, "PortPairGroupId Parent");
```

Comment:
we dont care about parent id. here again we are getting sfc directly

Reply:
This is delete network element. Which is we are deleting PPG from sfc. So network element will have element-id as PPG-id and parent-id will be the sfc id. So if see any of them null we have to throw an exception. Please let me know if i'm missing anything here.

Reply:
delete network element deletes the SFC. Similar to how register network elements creates a sfc.

To delete a PPG from an SFC, you can use the updateNetworkElements

Reply:
got it..fixed thanks